File: bindings/javascript/src/xmpp.js
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// XMPP protocol types for proven-servers.

/** StanzaType matching the Idris2 ABI tags. */
export const StanzaType = Object.freeze({
  MESSAGE: 0,
  PRESENCE: 1,
  IQ: 2,
});

/** MessageType matching the Idris2 ABI tags. */
export const MessageType = Object.freeze({
  CHAT: 0,
  MESSAGE_TYPE_ERROR: 1,
  GROUPCHAT: 2,
  HEADLINE: 3,
  NORMAL: 4,
});

/** PresenceType matching the Idris2 ABI tags. */
export const PresenceType = Object.freeze({
  AVAILABLE: 0,
  AWAY: 1,
  DND: 2,
  XA: 3,
  UNAVAILABLE: 4,
});

/** IqType matching the Idris2 ABI tags. */
export const IqType = Object.freeze({
  GET: 0,
  SET: 1,
  RESULT: 2,
  IQ_TYPE_ERROR: 3,
});

/** StreamError matching the Idris2 ABI tags. */
export const StreamError = Object.freeze({
  BAD_FORMAT: 0,
  CONFLICT: 1,
  CONNECTION_TIMEOUT: 2,
  HOST_GONE: 3,
  HOST_UNKNOWN: 4,
  NOT_AUTHORIZED: 5,
  POLICY_VIOLATION: 6,
  RESOURCE_CONSTRAINT: 7,
  SYSTEM_SHUTDOWN: 8,
});

File: bindings/dart/lib/src/xmpp.dart
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// XMPP protocol types for proven-servers.

/// StanzaType matching the Idris2 ABI tags.
enum StanzaType {
  message(0),
  presence(1),
  iq(2);

  const StanzaType(this.tag);
  final int tag;

  static StanzaType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// MessageType matching the Idris2 ABI tags.
enum MessageType {
  chat(0),
  messageType_Error(1),
  groupchat(2),
  headline(3),
  normal(4);

  const MessageType(this.tag);
  final int tag;

  static MessageType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// PresenceType matching the Idris2 ABI tags.
enum PresenceType {
  available(0),
  away(1),
  dnd(2),
  xa(3),
  unavailable(4);

  const PresenceType(this.tag);
  final int tag;

  static PresenceType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// IqType matching the Idris2 ABI tags.
enum IqType {
  get_(0),
  set_(1),
  result(2),
  iqType_Error(3);

  const IqType(this.tag);
  final int tag;

  static IqType? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// StreamError matching the Idris2 ABI tags.
enum StreamError {
  badFormat(0),
  conflict(1),
  connectionTimeout(2),
  hostGone(3),
  hostUnknown(4),
  notAuthorized(5),
  policyViolation(6),
  resourceConstraint(7),
  systemShutdown(8);

  const StreamError(this.tag);
  final int tag;

  static StreamError? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}
